File: shared/src/main/java/com/yuriy/openradio/broadcast/AbstractReceiver.java
package com.yuriy.openradio.broadcast;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import androidx.annotation.NonNull;

/**
 * Created by Chernyshov Yurii
 * At Android Studio
 * On 03/12/18
 * E-Mail: chernyshov.yuriy@gmail.com
 *
 * Base cass to handle broadcast receivers. For example, register and unregister safely.
 */
public abstract class AbstractReceiver extends BroadcastReceiver {

    private volatile boolean mIsRegistered;
    private final IntentFilter mIntentFilter;

    /**
     *
     * @param intentFilter
     */
    public AbstractReceiver(@NonNull final IntentFilter intentFilter) {
        super();
        mIsRegistered = false;
        mIntentFilter = intentFilter;
    }

    @Override
    public void onReceive(final Context context, final Intent intent) {
        /* Handle in implementor */
    }

    public void register(@NonNull final Context context) {
        if (!mIsRegistered) {
            context.registerReceiver(this, mIntentFilter);
        }
        mIsRegistered = true;
    }

    public void unregister(@NonNull final Context context) {
        if (mIsRegistered) {
            context.unregisterReceiver(this);
        }
        mIsRegistered = false;
    }
}

File: app/src/main/java/com/yuriy/openradio/shared/broadcast/AbstractReceiver.kt
package com.yuriy.openradio.shared.broadcast

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter

/**
 * Created by Chernyshov Yurii
 * At Android Studio
 * On 03/12/18
 * E-Mail: chernyshov.yuriy@gmail.com
 *
 * Base cass to handle broadcast receivers. For example, register and unregister safely.
 */
abstract class AbstractReceiver : BroadcastReceiver() {
    @Volatile
    private var mIsRegistered = false
    abstract fun makeIntentFilter(): IntentFilter?
    override fun onReceive(context: Context, intent: Intent) {
        /* Handle in implementor */
    }

    fun register(context: Context) {
        if (!mIsRegistered) {
            context.registerReceiver(this, makeIntentFilter())
        }
        mIsRegistered = true
    }

    open fun unregister(context: Context) {
        if (mIsRegistered) {
            context.unregisterReceiver(this)
        }
        mIsRegistered = false
    }
}